Source/Core/Common/Arm64Emitter.h

```diff
@@ -558,156 +534,44 @@ struct LogicalImm
-      // all the special case conditions in one go by seeing if c is zero.
-      if (a == 0)
-      {
-        // The input was zero (or all 1 bits, which will come to here too after we
-        // inverted it at the start of the function), which is invalid.
-        return;
-      }
-      else
-      {
-        // Otherwise, if c was zero but a was not, then there's just one stretch
-        // of set bits in our word, meaning that we have the trivial case of
-        // d == 64 and only one 'repetition'. Set up all the same variables as in
-        // the general case above, and set the N bit in the output.
-        clz_a = Common::CountLeadingZeros(a);
-        d = 64;
-        mask = ~UINT64_C(0);
-        out_n = 1;
-      }
-    }
-
-    // If the repeat period d is not a power of two, it can't be encoded.
-    if (!MathUtil::IsPow2<u64>(d))
+      valid = false;
       return;
+    }
 
-    // If the bit stretch (b - a) does not fit within the mask derived from the
-    // repeat period, then fail.
-    if (((b - a) & ~mask) != 0)
-      return;
+    // Normalize value, rotating it such that the LSB is 1:
+    // If LSB is already one, we mask away the trailing sequence of ones and
+    // pick the next sequence of ones. This ensures we get a complete element
+    // that has not been cut-in-half due to rotation across the word boundary.
 
-    // The only possible option is b - a repeated every d bits. Now we're going to
-    // actually construct the valid logical immediate derived from that
-    // specification, and see if it equals our original input.
-    //
-    // To repeat a value every d bits, we multiply it by a number of the form
-    // (1 + 2^d + 2^(2d) + ...), i.e. 0x0001000100010001 or similar. These can
-    // be derived using a table lookup on CLZ(d).
-    constexpr std::array<u64, 6> multipliers = {{
-        0x0000000000000001UL,
-        0x0000000100000001UL,
-        0x0001000100010001UL,
-        0x0101010101010101UL,
-        0x1111111111111111UL,
-        0x5555555555555555UL,
-    }};
-
-    const int multiplier_idx = Common::CountLeadingZeros((u64)d) - 57;
-
-    // Ensure that the index to the multipliers array is within bounds.
-    DEBUG_ASSERT((multiplier_idx >= 0) &&
-                 (static_cast<size_t>(multiplier_idx) < multipliers.size()));
-
-    const u64 multiplier = multipliers[multiplier_idx];
-    const u64 candidate = (b - a) * multiplier;
-
-    // The candidate pattern doesn't match our input value, so fail.
-    if (value != candidate)
-      return;
+    const size_t rotation = Common::CountTrailingZeros(value & (value + 1));
+    const u64 normalized = Common::RotateRight(value, rotation);
 
-    // We have a match! This is a valid logical immediate, so now we have to
-    // construct the bits and pieces of the instruction encoding that generates
-    // it.
-    n = out_n;
+    const size_t esize = Common::CountTrailingZeros(normalized & (normalized + 1));
```

Comment:
The meaning of `esize` wasn't obvious to me. Could we name this something like `element_size`? (Or is `esize` taken from some ARM document? If so, we can probably keep it as it is)

Reply:
esize is used as shortform for element size though most of the A64 manual; but renaming is fine.